### PR TITLE
ENH: Add estimator argument to gcm CI test #3027

### LINKED ISSUE #3027: [ENH] Add `estimator` argument to GCM conditional independence test
The current implementation of GCM test (pgmpy.estimators.CITests.gcm) uses a least square regression for computing residuals. To make the CI test more flexible, we should allow users to specify the estimator instead of using linear regression by default. We should add an `estimator` argument that accepts any scikit-learn estimator and then calls its `fit` and `predict` methods to compute the residuals.

### PR DESCRIPTION
**Pull request by Witchyass**
### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

Please answer the following questions:

- Did you use an LLM for any assistance with this PR? Please describe in **detail** (around a paragraph) how and what you used it for?  
[Yes, I used LLM assistance to understand edge cases during development. All code changes were applied and verified manually by me.]

- Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.  
[The gcm class was coded to use numpy's lstsq for computing residuals.
I replaced it with an estimator argument that accepts any scikit-learn regressor, defaulting to LinearRegression. 
I noticed that when Z=[] sklearn cannot fit on an empty feature set so I always append a column of ones to Z before fitting
 it is equivalent to the original intercept-only behavior]

- What steps have you taken to verify that the changes correctly address the issue? And what edge cases have you considered? Other than running tests, what else have you verified?  
[Ran existing test suite 
tested default estimator (LinearRegression)
tested custom estimator (RandomForestRegressor)
tested empty Z=[] 
tested invalid estimator (LogisticRegression)]

- Has the LLM added try-except blocks? They will need to be removed; any error handling must be explicit.  
[A try-except block was added to validate that the estimator is cloneable via sklearn.base.clone suggested by copilot ]

- Have you used LLM for generating tests? They need to be compressed into a smaller number of tests without reducing coverage.  
[No new tests were generated]

### Issue number(s) that this pull request fixes
- Fixes #3027

### List of changes to the codebase in this pull request
- ci_tests/gcm.py:
  - Added estimator=None to GCM init signature
  - Added imports (clone, LinearRegression)
  - Added estimator validation using hasattr
  - Replaced np.linalg.lstsq with sklearn fit/predict
  - Appended intercept column via np.column_stack
-tests/test_ci_tests/test_gcm.py:
 -Added RandomForestRegressor test
 -Updated expected test values to match current codebase
- estimators/CITests.py:
  - Reverted to original (changes moved to ci_tests/gcm.py)

### CODE REVIEW

**Review by ankurankan — CHANGES_REQUESTED** [review]
The new refactored CI tests are in `ci_tests` directory. Please make changes to that.

**Comment by Witchyass** [comment]
> The new refactored CI tests are in `ci_tests` directory. Please make changes to that.

Thanks for the clarification! I'll revert the changes in CITests.py and apply them to ci_tests/gcm.py instead

File: pgmpy/ci_tests/gcm.py [hunk 1]
```diff
@@ -9,15 +11,18 @@ class GCM(_BaseCITest):
     """
     The Generalized Covariance Measure(GCM) test for CI.
 
-    It performs linear regressions on the conditioning variable and then tests
-    for a vanishing covariance between the resulting residuals. Details of the
-    method can be found in [1].
+    It performs regressions on the conditioning variable using a specified
+    estimator and then tests for a vanishing covariance between the
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: using the term `regression` here can imply linear regression. This can be any estimator that works for continuous variables (regressor)

File: pgmpy/ci_tests/gcm.py [hunk 2]
```diff
@@ -9,15 +11,18 @@ class GCM(_BaseCITest):
...
+    resulting residuals. Details of the method can be found in [1].
 
     Parameters
     ----------
     data: pandas.DataFrame
         The dataset in which to test the independence condition.
-
+    estimator: sklearn estimator, optional (default=None)
+        A scikit-learn regressor to compute residuals. Must implement
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Same here. Can be any regressor with `fit` and `predict` methods; doesn't need to be from scikit-learn

File: pgmpy/ci_tests/gcm.py [hunk 3]
```diff
@@ -39,8 +44,14 @@ class GCM(_BaseCITest):
         "requires_data": True,
     }
 
-    def __init__(self, data: pd.DataFrame):
+    def __init__(self, data: pd.DataFrame, estimator=None):
         self.data = data
+        if estimator is None:
+            self.estimator = LinearRegression()
+        elif not is_regressor(estimator):
+            raise ValueError(f"estimator must be a scikit-learn regressor. Got {type(estimator)} instead.")
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: I think this check can cause problems when the estimator is not from scikit-learn. For example, xgboost. Maybe directly check if it has `fit` and `predict` methods.

File: pgmpy/ci_tests/gcm.py [hunk 4]
```diff
@@ -54,16 +65,18 @@ def run_test(
...
-        Z_aug = list(Z) + ["intercept"]
-        data_aug = data.assign(intercept=np.ones(data.shape[0]))
-
-        # Step 2: Compute the linear regression and the residuals
-        X_coef = np.linalg.lstsq(data_aug.loc[:, Z_aug], data_aug.loc[:, X], rcond=None)[0]
-        Y_coef = np.linalg.lstsq(data_aug.loc[:, Z_aug], data_aug.loc[:, Y], rcond=None)[0]
-        res_x = data_aug.loc[:, X] - data_aug.loc[:, Z_aug].dot(X_coef)
-        res_y = data_aug.loc[:, Y] - data_aug.loc[:, Z_aug].dot(Y_coef)
+        Z_data = data.loc[:, list(Z)].copy()
+        Z_data["__pgmpy_intercept__"] = np.ones(data.shape[0])
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Can call it simply `intercept`.

File: pgmpy/ci_tests/gcm.py [hunk 5]
```diff
@@ -54,16 +65,18 @@ def run_test(
...
         data = self.data
-        Z_aug = list(Z) + ["intercept"]
-        data_aug = data.assign(intercept=np.ones(data.shape[0]))
-
-        # Step 2: Compute the linear regression and the residuals
-        X_coef = np.linalg.lstsq(data_aug.loc[:, Z_aug], data_aug.loc[:, X], rcond=None)[0]
-        Y_coef = np.linalg.lstsq(data_aug.loc[:, Z_aug], data_aug.loc[:, Y], rcond=None)[0]
-        res_x = data_aug.loc[:, X] - data_aug.loc[:, Z_aug].dot(X_coef)
-        res_y = data_aug.loc[:, Y] - data_aug.loc[:, Z_aug].dot(Y_coef)
+        Z_data = data.loc[:, list(Z)].copy()
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Copy is a very expensive operation; try not to do this. Maybe numpy's column stack?

File: pgmpy/ci_tests/gcm.py
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Can replace `(1 - stats.norm.cdf(...)` with `stats.norm.sf(...)`. `sf` is more stable in certain cases.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by Witchyass** [comment]
Hi @ankurankan I noticed the Creating_Linear_BN.ipynb notebook test is failing but it seems unrelated to this PR
Could you confirm if this is a preexisting issue?

File: pgmpy/ci_tests/gcm.py [hunk 6]
```diff
@@ -12,6 +14,8 @@ class GCM(_BaseCITest):
     Regress :math:`X` and :math:`Y` on :math:`[1, Z]` using least squares, let :math:`r_X` and :math:`r_Y` denote the
     resulting residuals, and define :math:`U_i = r_{X, i} r_{Y, i}`. The resulting test statistic is
 
+    It fits a regressor on the conditioning variable and then tests for a vanishing covariance between the
+    resulting residuals. Details of the method can be found in [1].
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Could you please remove this sentence? I had updated the method description recently, and this is repetitive now.

File: pgmpy/ci_tests/gcm.py [hunk 7]
```diff
@@ -12,6 +14,8 @@ class GCM(_BaseCITest):
     Regress :math:`X` and :math:`Y` on :math:`[1, Z]` using least squares, let :math:`r_X` and :math:`r_Y` denote the
     resulting residuals, and define :math:`U_i = r_{X, i} r_{Y, i}`. The resulting test statistic is
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: Could you update this to say that we use an estimator instead of saying regress using least squares?

File: pgmpy/ci_tests/gcm.py [hunk 8]
```diff
@@ -58,16 +71,17 @@ def run_test(
 
         Sets ``self.statistic_`` (t-statistic) and ``self.p_value_``.
         """
-        # Step 1.1: Add another column with constant values to handle intercepts.
+        # Step 1.1: Append intercept column to ensure Z is never empty
```
Comment (ankurankan) [inline pgmpy/ci_tests/gcm.py]: `Step 1.1` -> `Step 1`

**Review by ankurankan — CHANGES_REQUESTED** [review]
@Witchyass Looks good mostly. I have added a few minor comments. Can you also update the tests `test_gcm.py` to test for some sklearn estimators.

**Comment by ankurankan** [comment]
> Hi @ankurankan I noticed the Creating_Linear_BN.ipynb notebook test is failing but it seems unrelated to this PR
Could you confirm if this is a preexisting issue?

Yes, we have an issue for it: https://github.com/pgmpy/pgmpy/issues/3128

**Comment by Witchyass** [comment]
> @Witchyass Looks good mostly. I have added a few minor comments. Can you also update the tests `test_gcm.py` to test for some sklearn estimators.

Hi @ankurankan  after the changes to gcm.py, the expected values in test_gcm no longer match. How would you like me to proceed?

File: pgmpy/ci_tests/gcm.py [hunk 9]
```diff
@@ -43,8 +50,14 @@ class GCM(_BaseCITest):
...
     }
 
-    def __init__(self, data: pd.DataFrame):
+    def __init__(self, data: pd.DataFrame, estimator=None):
         self.data = data
+        if estimator is None:
+            self.estimator = LinearRegression()
+        elif not (hasattr(estimator, "fit") and hasattr(estimator, "predict")):
+            raise ValueError(f"estimator must have fit and predict methods. Got {type(estimator)} instead.")
+        else:
```
[the remark this answers is not in the document]
Comment (ankurankan) [reply pgmpy/ci_tests/gcm.py:56]: I think this is a valid comment

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by ankurankan** [comment]
@Witchyass Can you investigate where exactly the difference is coming from? And how much is the difference?

**Comment by Witchyass** [comment]
Hi @ankurankan  I investigated the difference in expected test values in test_gcm.py

I ran the original lstsq approach on the same dataset 
(seed=42, n_samples=10000):
lstsq statistic: 39.631  (expected in test: 38.962)

The difference is not caused by our changes this suggests the expected values became outdated after a recent upstream change

The new values are now:
- Non-conditional statistic: 39.631, p_value: 0.0 (was 38.962, p_value: 0.0)
- Conditional independent statistic: 0.584, p_value: 0.5591 (was -0.312, p_value: 0.7547)
- Conditional dependent statistic: 39.798, p_value: 0.0 (was 39.798, p_value: 0.0)

**Comment by ankurankan** [comment]
@Witchyass The values look close enough; the difference seems to be coming from switching to linear regression. You can update the tests.

**Review by ankurankan — APPROVED** [review]
@Witchyass Thanks for the PR. Looks good to me. I have added a couple of minor changes:
1. Simplified the check for sklearn-compatible estimator.
2. Added an approx test suite as the exact one was getting skipped on github actions.
